src/lib/runway_takeoff/RunwayTakeoff.cpp

```diff
@@ -0,0 +1,286 @@
+		}
+
+		break;
+
+	case RunwayTakeoffState::CLAMPED_TO_RUNWAY:
+		if (airspeed > _airspeed_min.get() * _min_airspeed_scaling.get()) {
+			_state = RunwayTakeoffState::TAKEOFF;
+			mavlink_log_info(mavlink_fd, "#Takeoff airspeed reached");
+		}
+
+		break;
+
+	case RunwayTakeoffState::TAKEOFF:
+		if (alt_agl > _nav_alt.get()) {
+			_state = RunwayTakeoffState::CLIMBOUT;
+
+			/*
+			 * If we started in heading hold mode, move the navigation start WP to the current location now.
+			 * The navigator will take this as starting point to navigate towards the takeoff WP.
+			 */
+			if (_heading_mode.get() == 0) {
+				_start_wp(0) = (float)current_lat;
+				_start_wp(1) = (float)current_lon;
+			}
+
+			mavlink_log_info(mavlink_fd, "#Climbout");
+		}
+
+		break;
+
+	case RunwayTakeoffState::CLIMBOUT:
+		if (alt_agl > _climbout_diff.get()) {
+			_climbout = false;
+			_state = RunwayTakeoffState::FLY;
+			mavlink_log_info(mavlink_fd, "#Navigating to waypoint");
+		}
+
+		break;
+
+	default:
+		break;
+	}
+}
+
+/*
+ * Returns true as long as we're below navigation altitude
+ */
+bool RunwayTakeoff::controlYaw()
+{
+	// keep controlling yaw directly until we start navigation
+	return _state < RunwayTakeoffState::CLIMBOUT;
+}
+
+/*
+ * Returns pitch setpoint to use.
+ *
+ * Limited (parameter) as long as the plane is on runway. Otherwise
+ * use the one from TECS
+ */
+float RunwayTakeoff::getPitch(float tecsPitch)
+{
+	if (_state <= RunwayTakeoffState::CLAMPED_TO_RUNWAY) {
+		return math::radians(_runway_pitch_sp.get());
+	}
+
+	return tecsPitch;
+}
+
+/*
+ * Returns the roll setpoint to use.
+ */
+float RunwayTakeoff::getRoll(float navigatorRoll)
+{
+	// until we have enough ground clearance, set roll to 0
+	if (_state < RunwayTakeoffState::CLIMBOUT) {
+		return 0.0f;
+	}
+
+	// allow some roll during climbout
+	else if (_state < RunwayTakeoffState::FLY) {
+		return math::constrain(navigatorRoll,
+				       math::radians(-_max_takeoff_roll.get()),
+				       math::radians(_max_takeoff_roll.get()));
+	}
+
+	return navigatorRoll;
+}
+
+/*
+ * Returns the yaw setpoint to use.
+ *
+ * In heading hold mode (_heading_mode == 0), it returns initial yaw as long as it's on the
+ * runway. When it has enough ground clearance we start navigation towards WP.
+ */
+float RunwayTakeoff::getYaw(float navigatorYaw)
+{
+	if (_heading_mode.get() == 0 && _state < RunwayTakeoffState::CLIMBOUT) {
+		return _init_yaw;
+
+	} else {
+		return navigatorYaw;
+	}
+}
+
+/*
+ * Returns the throttle setpoint to use.
+ *
+ * Ramps up in the beginning, until it lifts off the runway it is set to
+ * parameter value, then it returns the TECS throttle.
+ */
+float RunwayTakeoff::getThrottle(float tecsThrottle)
+{
+	switch (_state) {
+	case RunwayTakeoffState::THROTTLE_RAMP: {
+			float throttle = hrt_elapsed_time(&_initialized_time) / (float)_throttle_ramp_time *
+					 _takeoff_throttle.get();
+			return throttle < _takeoff_throttle.get() ?
```

Comment:
I think this would be easier to understand if you just did:

math::constrain(throttle, 0, _takeoff_throttle.get())